source/adapters/level_zero/v2/common.hpp

```diff
@@ -104,5 +104,114 @@ using ze_context_handle_t =
+  }
+
+  ref_counted &operator=(ref_counted &&other) {
+    if (this == &other) {
+      return *this;
+    }
+
+    if (handle) {
+      release(handle);
+    }
+
+    handle = other.handle;
+    other.handle = nullptr;
+    return *this;
+  }
+
+  URHandle get() const { return handle; }
+
+private:
+  URHandle handle;
+};
+
+// This version of ref_counted does not call retain/release functions.
+// It is used to avoid circular references, most notably to ur_context_handle_t.
+// This is equivalent to just using URHandle but makes it clear that no ref
+// counting is expected.
+template <typename URHandle> struct weak {
+  template <ur_result_t (*retain)(URHandle), ur_result_t (*release)(URHandle)>
+  weak(const ref_counted<URHandle, retain, release> &handle)
+      : handle(handle.get()) {}
+
+  operator URHandle() const { return handle; }
+  URHandle operator->() const { return handle; }
+
+  weak(const weak &) = default;
+  weak &operator=(const weak &) = default;
+
+  weak(weak &&other) = default;
+  weak &operator=(weak &&other) = default;
+
+  URHandle get() const { return handle; }
+
+private:
+  URHandle handle;
+};
+
+template <typename URHandle> struct ref_counted_traits;
+
+#define DECLARE_REF_COUNTER_TRAITS(URHandle, retainFn, releaseFn)              \
+  template <> struct ref_counted_traits<URHandle> {                            \
+    static ur_result_t retain(URHandle handle) {                               \
+      assert(handle);                                                          \
+      return retainFn(handle);                                                 \
+    }                                                                          \
+    static ur_result_t release(URHandle handle) {                              \
+      assert(handle);                                                          \
+      return releaseFn(handle);                                                \
+    }                                                                          \
+    static ur_result_t validate([[maybe_unused]] URHandle handle) {            \
+      assert(handle);                                                          \
+      assert(reinterpret_cast<_ur_object *>(handle)->RefCount.load() != 0);    \
+      return UR_RESULT_SUCCESS;                                                \
+    }                                                                          \
+  };
+
+// This version of ref_counted calls retain/release functions.
+template <typename URHandle>
+using rc = ref_counted<URHandle, ref_counted_traits<URHandle>::retain,
+                       ref_counted_traits<URHandle>::release>;
+
+// This version of ref_counted validates that the ref count is not zero on every
+// release and retain in debug mode, and does nothing in the release mode.
+// Used for types that should always be alibe during the adapter lifetime (e.g.
+// devices).
+template <typename URHandle>
+using rc_val_only =
```

Comment:
hmm, I don't like the name :D

`durable`, `NonNull`, `Live`, hm, dunno. Basically what we want to say is that this object's lifetime is guaranteed to exceed that of the containing object.

Reply:
Yeah, me too. `non_null`/`not_null` might good, or maybe `checked`?

Reply:
i like `checked`.